api/src/opentrons/protocol_engine/state/pipettes.py

```
@@ -107,10 +109,20 @@ def _handle_command(self, command: Command) -> None:
         elif isinstance(command.result, DispenseResult):
             pipette_id = command.params.pipetteId
             previous_volume = self._state.aspirated_volume_by_id[pipette_id]
-            next_volume = max(0, previous_volume - command.result.volume)
-
+            next_volume = max(0.0, previous_volume - command.result.volume)
             self._state.aspirated_volume_by_id[pipette_id] = next_volume
 
+        elif isinstance(command.result, PickUpTipResult):
+            pipette_id = command.params.pipetteId
+            tiprack_id = command.params.labwareId
+            self._state.attached_tip_labware_by_id[pipette_id] = tiprack_id
+
+        elif isinstance(command.result, DropTipResult):
+            pipette_id = command.params.pipetteId
+            if self._state.attached_tip_labware_by_id.get(pipette_id):
+                self._state.attached_tip_labware_by_id.pop(pipette_id)
+            # TODO (spp): If no pipette-id, log warning instead of just no-oping?
```

`pipette_id` should never be missing from `self._state.attached_tip_labware_by_id` unless there's a bug in our code, right?


If that's right, IMO I would just do an unguarded `.pop(pipette_id)` and let it raise `KeyError`.

---

I went back and forth on this. Having it no-op if `pipette_id` is not found makes testing easier. Otherwise every time a drop tip command is used in a test, we will have to make sure that `attached_tip_labware_by_id` has an entry with the `pipette_id`. An example is the parametrized test `test_movement_commands_update_current_well` which takes in individual commands as parameters and checks if `current_well` updated correctly upon command handling.
But if we decide that it is important to have it raise if no `pipette_id` found, then yes the tests should reflect that too and will need to be modified. 
Decided on no-oping with maybe a warning because we don't gain anything by raising the error. The state update intends to remove the `pipette_id` entry from `attached_tip_labware_by_id` anyway so there's no harm done if the entry didn't exist in the first place. If id itself was wrong or missing then it'll be caught way before this line is executed anyway.